gloo/transport/uv/unbound_buffer.cc

```diff
@@ -0,0 +1,133 @@
+/**
+ * Copyright (c) 2019-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <gloo/transport/uv/unbound_buffer.h>
+
+#include <stdexcept>
+
+#include <gloo/common/error.h>
+#include <gloo/common/logging.h>
+#include <gloo/transport/uv/context.h>
+
+namespace gloo {
+namespace transport {
+namespace uv {
+
+UnboundBuffer::UnboundBuffer(
+    std::shared_ptr<Context> context,
+    void* ptr,
+    size_t size)
+    : ::gloo::transport::UnboundBuffer(ptr, size),
+      context_(context),
+      recvCompletions_(0),
+      recvRank_(-1),
+      sendCompletions_(0),
+      sendRank_(-1),
+      shareableNonOwningPtr_(this) {}
+
+UnboundBuffer::~UnboundBuffer() {}
+
+void UnboundBuffer::handleRecvCompletion(int rank) {
+  std::lock_guard<std::mutex> lock(mutex_);
+  recvCompletions_++;
+  recvRank_ = rank;
+  recvCv_.notify_one();
+}
+
+void UnboundBuffer::waitRecv(int* rank, std::chrono::milliseconds timeout) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  if (timeout == kUnsetTimeout) {
+    timeout = context_->getTimeout();
+  }
+
+  if (recvCompletions_ == 0) {
+    auto done =
+        recvCv_.wait_for(lock, timeout, [&] { return recvCompletions_ > 0; });
+    if (!done) {
+      throw ::gloo::IoException(GLOO_ERROR_MSG(
+          "Timed out waiting ",
+          timeout.count(),
+          "ms for recv operation to complete"));
+    }
+  }
+
+  recvCompletions_--;
+  if (rank != nullptr) {
+    *rank = recvRank_;
+  }
+}
+
+void UnboundBuffer::handleSendCompletion(int rank) {
+  std::lock_guard<std::mutex> lock(mutex_);
+  sendCompletions_++;
+  sendRank_ = rank;
+  sendCv_.notify_one();
+}
+
+void UnboundBuffer::waitSend(int* rank, std::chrono::milliseconds timeout) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  if (timeout == kUnsetTimeout) {
+    timeout = context_->getTimeout();
+  }
+
+  if (sendCompletions_ == 0) {
+    auto done =
+        sendCv_.wait_for(lock, timeout, [&] { return sendCompletions_ > 0; });
+    if (!done) {
+      throw ::gloo::IoException(GLOO_ERROR_MSG(
+          "Timed out waiting ",
+          timeout.count(),
+          "ms for send operation to complete"));
+    }
+  }
+
+  sendCompletions_--;
+  if (rank != nullptr) {
+    *rank = sendRank_;
+  }
+}
+
+void UnboundBuffer::send(
+    int dstRank,
+    uint64_t slot,
+    size_t offset,
+    size_t nbytes) {
+  if (nbytes == UINT64_MAX) {
```

Comment:
It does. But for clarity it should use `SIZE_MAX` or `std::numeric_limits<size_t>::max()` because we're dealing with `size_t` here. Will address and change this across the board.

Reply:
I added #200 to my stack of diffs that addresses this.

I plan to rebase this PR once that stack is merged. They're all preparation diffs for compiling on macOS.